### PR TITLE
[WFTC-136] Memory leak, remove XAResourceRecovery when reload

### PR DESCRIPTION
**Pull request by marcosgopen**
https://issues.redhat.com/browse/WFTC-136

### CODE REVIEW

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java [hunk 1]
```diff
@@ -94,7 +96,13 @@ public abstract class JBossLocalTransactionProvider implements LocalTransactionP
         }
         this.fileSystemXAResourceRegistry = new FileSystemXAResourceRegistry(this, xaRecoveryDirRelativeToPath);
         XAResourceRegistryProviderHolder.register(fileSystemXAResourceRegistry::getInDoubtXAResources);
-        registry.addXAResourceRecovery(fileSystemXAResourceRegistry::getInDoubtXAResources);
+        xaResourceRecovery = fileSystemXAResourceRegistry::getInDoubtXAResources;
```
Comment (mmusgrov) [inline src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:99]: For clarity, how about moving this before line 98 and then replace line 98 with `XAResourceRegistryProviderHolder.register(xaResourceRecovery)`?

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java [hunk 2]
```diff
@@ -94,7 +96,13 @@ public abstract class JBossLocalTransactionProvider implements LocalTransactionP
...
+        registry.addXAResourceRecovery(xaResourceRecovery);
+    }
+
+    public void removeXAResourceRecovery(XAResourceRecoveryRegistry registry) {
+        registry.removeXAResourceRecovery(xaResourceRecovery);
+
```
Comment (mmusgrov) [inline src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:105]: Superfluous blank line.

**Review by mmusgrov — APPROVED** [review]
Yes this is a nice fix and I definitely approve of it!

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java (hunk 1 above)
Comment (marcosgopen) [reply src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:99]: I had tought about it. But actually the functional interface for XAResourceRegistryProviderHolder.register() is 'XAResourceRegistryProvider' while the functional interface for registry.addXAResourceRecovery() is 'XAResourceRecovery'. So using the same reference is not allowed.

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java (hunk 2 above)
Comment (marcosgopen) [reply src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:105]: Thanks!

Comment (marcosgopen) [reply src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java]: Fixed

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java [hunk 3]
```diff
@@ -77,6 +78,7 @@ public abstract class JBossLocalTransactionProvider implements LocalTransactionP
     private final ConcurrentSkipListSet<XidKey> timeoutSet = new ConcurrentSkipListSet<>();
     private final ConcurrentMap<SimpleXid, Entry> known = new ConcurrentHashMap<>();
     private final FileSystemXAResourceRegistry fileSystemXAResourceRegistry;
+    private XAResourceRecovery xaResourceRecovery;
```
Comment (chengfang) [inline src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:81]: this field can be final.

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java [hunk 4]
```diff
@@ -94,7 +96,12 @@ public abstract class JBossLocalTransactionProvider implements LocalTransactionP
...
         this.fileSystemXAResourceRegistry = new FileSystemXAResourceRegistry(this, xaRecoveryDirRelativeToPath);
         XAResourceRegistryProviderHolder.register(fileSystemXAResourceRegistry::getInDoubtXAResources);
-        registry.addXAResourceRecovery(fileSystemXAResourceRegistry::getInDoubtXAResources);
+        xaResourceRecovery = fileSystemXAResourceRegistry::getInDoubtXAResources;
+        registry.addXAResourceRecovery(xaResourceRecovery);
+    }
+
+    public void removeXAResourceRecovery(XAResourceRecoveryRegistry registry) {
+        registry.removeXAResourceRecovery(xaResourceRecovery);
     }
```
Comment (chengfang) [inline src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:105]: Can we add javadoc to this method, including how it's used in consumer project like WildFly?

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java (hunk 3 above)
Comment (marcosgopen) [reply src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java]: Thanks @chengfang for your review! I addressed your comment.

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java (hunk 1 above)
Comment (mmusgrov) [reply src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:99]: Under the covers the runtime must know how to convert an xaResourceRecovery instance into an array of them.
Thanks for the clarification.

File: src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java (hunk 4 above)
Comment (marcosgopen) [reply src/main/java/org/wildfly/transaction/client/provider/jboss/JBossLocalTransactionProvider.java:113]: Thanks @chengfang , I added the javadoc, WDYT?

**Review by chengfang — APPROVED** [review]

**Comment by chengfang** [comment]
all 6 checks passed.

**Comment by chengfang** [comment]
@marcosgopen @mmusgrov thanks!

**Pull request merged**